README.md
# owntracks-cli-publisher

![OCLI logo](assets/owntrackscli192.png)

This is the OwnTracks command line interface publisher, a.k.a. _owntracks-cli-publisher_, a small utility which connects to _gpsd_ and publishes position information in [OwnTracks JSON](https://owntracks.org/booklet/tech/json/) to an MQTT broker in order for [compatible software](https://owntracks.org/booklet/guide/clients/) to process location data. (Read up on [what OwnTracks does](https://owntracks.org/booklet/guide/whathow/) if you're new to it.)

### gpsd

[gpsd] is a daemon that receives data from one or more GPS receivers and provides data to multiple applications via a TCP/IP service. GPS data thus obtained is processed by _owntracks-cli-publisher_ and published via [MQTT].

![vk-172](assets/img_9643.jpg)

We've been successful with a number of GPS receivers, some of which are very affordable. The small USB stick receiver in the above photo is called a VK172 and costs roughly €16.

Determining which USB port was obtained by the receiver can be challenging, but _dmesg_ is useful to find out. Once you know the port, cross your fingers that it remains identical after a reboot, and ensure your _gpsd_ service is running. For testing, we've found that launching it from a terminal is useful:

```console
$ gpsd -n -D 2 -N /dev/tty.usbmodem1A121201
```
### owntracks-cli-publisher

_owntracks-cli-publisher_ operates with a number of defaults which you can override using environment variables.

The following defaults are used:

- The MQTT base topic defaults to `owntracks/<username>/<hostname>`, where _username_ is the name of the logged in user, and _hostname_ the short host name. This base topic can be overridden by setting `BASE_TOPIC` in the environment.
- The MQTT host and port are `localhost` and `1883` respectively, and can be set using `MQTT_HOST` and `MQTT_PORT`.
- The MQTT clientId is set to `"ocli-<username>-<hostname>"`, but it can be overridden by setting `OCLI_CLIENTID` in the environment.
- TCP is used to connect to _gpsd_ with `localhost` and `2947` being the default host and port, overridden by setting `GPSD_HOST` and `GPSD_PORT`.
- The two-letter OwnTracks [tracker ID](https://owntracks.org/booklet/features/tid/) can be configured by setting `OCLI_TID`; it defaults to not being used.
- `OCLI_INTERVAL` defaults to 60 seconds.
- `OCLI_DISPLACEMENT` defaults to 0 meters.
- TLS can be enabled for the MQTT connection by specifying the path to a PEM CA certificate with which to verify peers in `OCLI_CACERT`. Note, that you'll likely need to also specify a different `MQTT_PORT` from the default.

_owntracks-cli-publisher_ reads GPS data from _gpsd_ and as soon as it has a fix it publishes an OwnTracks payload (see below). _owntracks-cli-publisher_ will subsequently publish a message every `OCLI_INTERVAL` seconds or when it detects it has moved `OCLI_DISPLACEMENT` meters.

![owntracks-cli-publisher with OwnTracks on macOS](assets/jmbp-5862.png)

#### payload

Any number of path names can be passed as arguments to _owntracks-cli-publisher_ which interprets each in terms of an element which will be added to the OwnTracks JSON. The element name is the base name of the path. If a path points to an executable file the first line of _stdout_ produced by that executable will be used as the _key_'s _value_, otherwise the first line read from the file. In both cases, trailing newlines are removed from values.

```console
$ echo 27.2 > parms/temp
$ owntracks-cli-publisher parms/temp contrib/platform
```

In this example, we use a file and a program. When _owntracks-cli-publisher_ produces its JSON we'll see something like this:

```json
{
  "_type": "location",
  "tst": 1577654651,
  "lat": 48.856826,

  "temp" : "27.2",
  "platform": "FreeBSD"
}
```

Note that a _key_ may not overwrite JSON keys defined by _owntracks-cli-publisher_, so for example, a file called `lat` will not be accepted as it would clobber the latitude JSON element.

#### controlling owntracks-cli-publisher

It is possible to control _owntracks-cli-publisher_ using a subset of OwnTrack's `cmd` commands.

```console
$ t=owntracks/jpm/tiggr/cmd
$ mosquitto_pub -t $t -m "$(jo _type=cmd action=reportLocation)"
```
The following commands are currently implemented:

- `reportLocation` causes _owntracks-cli-publisher_ to publish its current location (providing _gpsd_ has a fix). _owntracks-cli-publisher_ sets `t:m` in the JSON indicating the publish was manually requested.
- `dump` causes _owntracks-cli-publisher_ to publish its internal configuration to the topic `<basetopic>/dump` as a `_type: configuration` message.

	```json
	{
	  "_type": "configuration",
	  "_npubs": 47,
	  "clientId": "owntracks-ocli",
	  "locatorInterval": 60,
	  "locatorDisplacement": 0,
	  "pubTopicBase": "owntracks/jpm/tiggr",
	  "tid": "OC",
	  "username": "jpm",
	  "deviceId": "tiggr"
	}
	```

- `setConfiguration` permits setting some of _owntracks-cli-publisher_'s internal values. Note that these do not persist a restart.

    ```console
    $ mosquitto_pub -t $t -m "$(jo _type=cmd action=setConfiguration configuration=$(jo _type=configuration locatorInterval=10 locatorDisplacement=0))"
    ```

	```json
	{
	  "_type": "cmd",
	  "action": "setConfiguration",
	  "configuration": {
	    "_type": "configuration",
	    "locatorInterval": 10,
	    "locatorDisplacement": 0
	  }
	}
	```



### testing

There is a small set of scripts with which you can test owntracks-cli-publisher without having a real GPS receiver. Please check [contrib/fake/](contrib/fake/) for more information.

### building

_owntracks-cli-publisher_ should compile easily once you extract the source code and have the prerequisite libraries installed for linking against _gpsd_ and the _mosquitto_ library.

Systems we've tested on require the following packages in order to build _owntracks-cli-publisher_.

#### FreeBSD

```console
# pkg install mosquitto gpsd
$ make
```

#### OpenBSD

```console
# pkg_add mosquitto gpsd
$ make
```

#### macOS

```console
$ brew install mosquitto gpsd
$ make
```

#### Debian

```console
# apt-get install libmosquitto-dev libgps-dev
$ make
```

#### CentOS

```console
# yum install openssl-devel
$ wget https://mosquitto.org/files/source/mosquitto-1.6.8.tar.gz
$ tar xf mosquitto-1.6.8.tar.gz
$ cd mosquitto-1.6.8/
$ make
# make install
# echo /usr/local/lib > /etc/ld.so.conf.d/mosquitto.conf
# ldconfig

# yum install python3-scons
$ wget http://download-mirror.savannah.gnu.org/releases/gpsd/gpsd-3.19.tar.gz
$ tar xf gpsd-3.19.tar.gz
$ cd gpsd-3.19/
$ /usr/bin/scons-3
# /usr/bin/scons-3 udev-install

$ cd ocli
$ make
```

#### systemd

This may be a way of getting _owntracks-cli-publisher_ working on machines with _systemd_. Basically we need two things:

1. an environment file, `owntracks-cli-publisher.env`:

	```
	name="Testing"
	fixlog="/tmp/fix.log"
	BASE_TOPIC="m/bus/b001"
	MQTT_HOST="localhost"
	MQTT_PORT=1888
	GPSD_HOST="localhost"
	GPSD_PORT="2947"
	OCLI_TID="OC"
	```

2. a systemd Unit file:

	```
	[Unit]
	Description=OwnTracks cli
	Requires=mosquitto.service

	[Service]
	Type=simple
	EnvironmentFile=/home/jpm/owntracks-cli-publisher.env
	ExecStartPre=/usr/bin/touch /tmp/ocli-started-${name}
	ExecStart=/home/jpm/bin/owntracks-cli-publisher
	Restart=always
	RestartSec=60
	User=mosquitto
	Group=mosquitto

	[Install]
	WantedBy=multi-user.target
	```

### Packages

Packages are available for select operating systems from the Github releases page. Note that the RPM packages do not contain an `owntracks-cli-publisher.env` file for systemd.


### Credits

- Idea and initial implementation by [Jan-Piet Mens](https://jpmens.net)
- [gpsd]
- [mosquitto](https://mosquitto.org)
- [utarray](https://troydhanson.github.io/uthash/utarray.html)
- [utstring](https://troydhanson.github.io/uthash/utstring.html)
- packaging work by [Andreas Motl](https://github.com/amotl)

  [gpsd]: https://gpsd.gitlab.io/gpsd/
  [mqtt]: http://mqtt.org


CHANGES.md
# ocli changelog

## 2020-02-15 0.9.1
- Add package building for CentOS 6,7,8
- Rename package to "owntracks-cli-publisher"
- Use version tag from "version.h".
- silence switch is honoured

## 2020-01-12 0.8.0
- add support for pledge(2) and unveil(2) in OpenBSD
- Improve fake gps data for Paris, thanks to @ckrey
- Clarify subscription
- Add version, adapt CFLAGS, and -s and -v
- Add infrastructure for Docker-based packaging (#4)
- Switch to Semantic Versioning
- Decrease default publish interval to 1s
- Add fake data generator
- Replace `bash` markdown code blocks with `console`; by @linusg
- Initial import from jps2m
- Add support for file params
- Support for reading from executable params
- Add -I and -L for *BSD
- Bring to GPSD API level 8 (for OpenBSD)
- Started on documentation
- Carry mosq pointer in udata
- Put all configurables in udata structure
- Document parms
- Rename minsecs and minmove consistently
- Document building; lol centos8
- Add some support for remoteConfig
- Make clientid configurable
- OwnTracks on macOS
- Add license file
- Change executable uname to contrib/platform
- Thank you, @ckrey, for the logo!
- Change default MQTT clientid
- Dump crashes if tid unset; reported by @ckrey
- reportLocation should print now() in tst; reported by @ckrey
- Publish all locations with retain; reported by @ckrey
- Selectively use retain; cleanup
- Subscribe on connect so we subscribe after auto-reconnect
- [platform] Be more specific for FreeBSD
- Add support for TLS; general cleanup; closes #1


ocli.c
/*
 * OwnTracks CLI
 * Copyright (C) 2016-2020 Jan-Piet Mens <jpmens@gmail.com>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>
#include <syslog.h>
#include <math.h>
#include <time.h>
#include <errno.h>
#include <libgen.h>
#include <signal.h>
#include <assert.h>
#include <unistd.h>
#include <gps.h>
#include <libgen.h>
#include <getopt.h>
#include <mosquitto.h>
#include "json.h"
#include "utarray.h"
#include "utstring.h"
#include "version.h"

#define BLEN		8192

#define QOS0		0
#define QOS1		1
#define QOS2		2
#define SSL_VERIFY_PEER (1)
#define SSL_VERIFY_NONE (0)

#define max(a, b) ( (a > b) ? a : b )

#define PERIODIC_REPORT	NULL
#define ONDEMAND_REPORT "m"

/* Globals, configured from environment */
FILE *fixlog = NULL;		/* if open, record JSON to this file */

UT_array *parms;
static struct gps_data_t gpsdata;
long npubs = 0L;
#if GPSD_API_MAJOR_VERSION >= 7
char gpsmessage[BLEN];
size_t gpsmessagelen = BLEN;
#endif

#define PROGNAME "ocli"
#define SIESTA	700		/* microseconds */

struct udata {
	struct mosquitto *mosq;
	char *clientid;
	char *basetopic;
	char *t_dump;		/* topic on which to dump configuration to */
	char *t_cmd;		/* topic on which to subscribe to for commands */
	char *tid;
	char *username;
	char *device;

	int interval;		/* publish after seconds */
	int displacement;	/* publish after this number of meters movement (def: 0) */
	bool verbose;
};

void publish(struct udata *ud, char *topic, char *payload, int qos, bool retain);
static void print_fix(struct udata *ud, struct gps_data_t *gpsdata, double time, char *reporttype);

void catcher(int sig)
{
	fprintf(stderr, "%s: Going down on signal %d", PROGNAME, sig);
	exit(1);
}

void publish(struct udata *ud, char *topic, char *payload, int qos, bool retain)
{
	int rc;

	rc = mosquitto_publish(ud->mosq, NULL, topic, strlen(payload), payload, qos, retain);
	if (rc != MOSQ_ERR_SUCCESS) {
		fprintf(stderr, "%s: cannot publish: %s\n", PROGNAME, mosquitto_strerror(rc));
		exit(1);
	}
}

void cb_connect(struct mosquitto *mosq, void *userdata, int reason)
{
	struct udata *ud = (struct udata *)userdata;
	int rc;

	if ((rc = mosquitto_subscribe(ud->mosq, NULL, ud->t_cmd, QOS1)) != MOSQ_ERR_SUCCESS) {
		fprintf(stderr, "cannot subscribe to %s: %s\n", ud->t_cmd,
			mosquitto_strerror(rc));
	}
}

void cb_disconnect(struct mosquitto *mosq, void *userdata, int reason)
{
	if (reason == 0) {
		// Disconnect requested by client
	} else {
		fprintf(stderr, "%s: disconnected: reason: %d (%s)\n",
			PROGNAME, reason, strerror(errno));
		sleep(2);
	}
}

static void config_dump(struct udata *ud)
{
	char *json_string;
	JsonNode *jo;

	jo = json_mkobject();

	json_append_member(jo, "_type",			json_mkstring("configuration"));
	json_append_member(jo, "_npubs",		json_mknumber(npubs));		// nonstandard
	json_append_member(jo, "clientId", 		json_mkstring(ud->clientid));
	json_append_member(jo, "locatorInterval",	json_mknumber(ud->interval));
	json_append_member(jo, "locatorDisplacement",	json_mknumber(ud->displacement));
	json_append_member(jo, "pubTopicBase",		json_mkstring(ud->basetopic));
	if (ud->tid && *ud->tid) {
		json_append_member(jo, "tid", 			json_mkstring(ud->tid));
	}
	json_append_member(jo, "username", 		json_mkstring(ud->username));
	json_append_member(jo, "deviceId", 		json_mkstring(ud->device));

	if ((json_string = json_stringify(jo, NULL)) != NULL) {
		publish(ud, ud->t_dump, json_string, QOS1, false);

		free(json_string);
	}

	json_delete(jo);
}

void set_config(struct udata *ud, JsonNode *json, char *payload)
{
	JsonNode *conf, *j;

	/* FIXME: add checking for remoteConfiguration to permit changes */

	if ((conf = json_find_member(json, "configuration")) == NULL) {
		fprintf(stderr, "No configuration in JSON %s\n", (char *)payload);
		return;
	}

	/*
	 * We now have:
	 * {
	 *     "_type": "configuration",
	 *     "bla": "foo"
	 * }
	 */

	if (((j = json_find_member(conf, "_type")) == NULL) ||
		(strcmp(j->string_, "configuration") != 0)) {
		fprintf(stderr, "No configuration in config action %s\n", payload);
		return;
	}

	json_foreach(j, conf) {
		if (strcmp(j->key, "_type") == 0)
			continue;
		// printf("%s\t%s\n", j->key, j->string_);

		if (strcmp(j->key, "locatorInterval") == 0) {
			ud->interval = j->number_;
			fprintf(stderr, "Set interval to %d\n", ud->interval);
		} else if (strcmp(j->key, "locatorDisplacement") == 0) {
			ud->displacement = j->number_;
			fprintf(stderr, "Set displacement to %d\n", ud->displacement);
		} else {
			fprintf(stderr, "Ignoring unknown configuration key `%s'\n", j->key);
		}
	}
}

void cb_message(struct mosquitto *mosq, void *userdata, const struct mosquitto_message *msg)
{
	struct udata *ud = (struct udata *)userdata;
	JsonNode *json, *j;
	char *action = NULL;

	if (msg->payloadlen < 2)
		return;

	if ((json = json_decode((char *)msg->payload)) == NULL) {
		fprintf(stderr, "Cannot decode JSON from %s\n", (char *)msg->payload);
		return;
	}

	if (((j = json_find_member(json, "_type")) == NULL) ||
		(strcmp(j->string_, "cmd") != 0)) {
		fprintf(stderr, "No cmd in JSON %s\n", (char *)msg->payload);
		json_delete(json);
		return;
	}

	if ((j = json_find_member(json, "action")) == NULL) {
		fprintf(stderr, "No action in JSON %s\n", (char *)msg->payload);
		json_delete(json);
		return;
	}

	if (j->tag != JSON_STRING) {
		fprintf(stderr, "action tag is not string in JSON %s\n", (char *)msg->payload);
		json_delete(json);
		return;
	}

	action = strdup(j->string_);

	if (strcmp(action, "dump") == 0) {
		config_dump(ud);
	} else if (strcmp(action, "reportLocation") == 0) {
		print_fix(ud, &gpsdata, (time_t)time(0), ONDEMAND_REPORT);
	} else if (strcmp(action, "setConfiguration") == 0) {
		set_config(ud, json, (char *)msg->payload);
	}

	json_delete(json);
	if (action)
		free(action);
}

static void print_fix(struct udata *ud, struct gps_data_t *gpsdata, double ttime, char *reporttype)
{
	char buf[128], *json_string, **p;
	double accuracy;
	struct gps_fix_t *fix = &(gpsdata->fix);
	JsonNode *jo;
	char tbuf[128];

	if (isnan(fix->latitude) ||
		isnan(fix->longitude) ||
		isnan(ttime)) {
		return;
	}

	if (!isnan(fix->epx) && !isnan(fix->epy)) {
		accuracy = max(fix->epx, fix->epy);
	} else if (isnan(fix->epx) && !isnan(fix->epy)) {
		accuracy = fix->epy;
	} else if (!isnan(fix->epx) && isnan(fix->epx)) {
		accuracy = fix->epx;
	} else {
		accuracy = 0.0;
	}

	unix_to_iso8601(ttime, tbuf, sizeof(tbuf));
	if (ud->verbose) {
		printf("mode=%d, lat=%f, lon=%f, acc=%f, tst=%s (%ld)\n",
			fix->mode,
			fix->latitude,
			fix->longitude,
			accuracy,
			tbuf,
			(long)ttime);
	}

	jo = json_mkobject();

	json_append_member(jo, "_type",	json_mkstring("location"));
	json_append_member(jo, "lat",	json_mknumber(fix->latitude));
	json_append_member(jo, "lon",	json_mknumber(fix->longitude));
	json_append_member(jo, "acc",	json_mknumber((long)accuracy));
	json_append_member(jo, "tst",	json_mknumber(ttime));

	if (ud->tid && *ud->tid) {
		json_append_member(jo, "tid", json_mkstring(ud->tid));
	}

	if (reporttype) {
		json_append_member(jo, "t", json_mkstring(reporttype));
	}

	/* DEBUGGING */
	sprintf(buf, "%d/%d", gpsdata->satellites_used, gpsdata->satellites_visible);
	json_append_member(jo, "sat", json_mkstring(buf));
	if ((fix->mode == MODE_3D) && !isnan(fix->altitude)) {
		json_append_member(jo, "alt", json_mknumber((long)fix->altitude));
	}
	if (!isnan(fix->speed)) {
		long kph = (fix->speed * 3.6); 	/* meters/s -> km/h */
		json_append_member(jo, "vel", json_mknumber(kph));
	}

	/*
	 * For each of the filenames passed as parameters, use the basename
	 * as key for the string value of the first line in the file and add
	 * to the JSON. Skip if the key is already in the JSON object, i.e.
	 * don't overwrite existing elements.
	 */

	p = NULL;
	while ((p = (char**)utarray_next(parms, p))) {
		char *key = basename(*p), *val = NULL;
		FILE *fp;
		bool is_exec = false;

		if (json_find_member(jo, key) != NULL) {
			// fprintf(stderr, "Refuse to overwrite key=%s\n", key);
			continue;
		}

		is_exec = access(*p, X_OK) == 0;

		if (is_exec) {
			fp = popen(*p, "r");
			is_exec = true;
		} else {
			fp = fopen(*p, "r");
		}

		if (fp == NULL) {
			perror(*p);
			continue;
		} else {
			char buf[1025], *bp;
			if (fgets(buf, sizeof(buf), fp) != NULL) {
				if ((bp = strchr(buf, '\r')) != NULL)
					*bp = 0;
				if ((bp = strchr(buf, '\n')) != NULL)
					*bp = 0;
				val = buf;
			}
			if (is_exec) {
				pclose(fp);
			} else {
				fclose(fp);
			}
		}
		if (val) {
			json_append_member(jo, key, json_mkstring(val));
		} else {
			json_append_member(jo, key, json_mknull());
		}
	}

	if ((json_string = json_stringify(jo, NULL)) != NULL) {
		npubs++;
		publish(ud, ud->basetopic, json_string, QOS1, true);

		if (fixlog) {
			fprintf(fixlog, "%s\n", json_string);
			fflush(fixlog);	/* FIXME */
		}
		free(json_string);
	}

	json_delete(jo);
}


static void conditionally_log_fix(struct udata *ud, struct gps_data_t *gpsdata)
{
	struct gps_fix_t *fix = &(gpsdata->fix);
	static double int_time, old_int_time;
	static double old_lat, old_lon;
	static bool first = true;
	bool valid = false;

	if (gpsdata->set & POLICY_SET) {
		gpsdata->set &= ~(POLICY_SET);
		return;
	}

	if (gpsdata->set & STATUS_SET) {
		switch (gpsdata->status) {
			case STATUS_FIX:
#ifdef STATUS_DGPS_FIX
			case STATUS_DGPS_FIX:
#endif
				switch (gpsdata->fix.mode) {
					case MODE_2D:
						if (gpsdata->set & LATLON_SET) {
							valid = true;
						}
						break;

					case MODE_3D:
						if (gpsdata->set & (LATLON_SET|ALTITUDE_SET)) {
							valid = true;
						}
						break;

					case MODE_NOT_SEEN:
						if (ud->verbose) {
							fprintf(stderr, ".. fix not yet seen\n");
						}
						break;

					case MODE_NO_FIX:
						if (ud->verbose) {
							fprintf(stderr, ".. no fix yet\n");
						}
						break;

					default:
						if (ud->verbose) {
							fprintf(stderr, ".. unpossible mode\n");
						}
						break;
				}
				break;

			case STATUS_NO_FIX:
				if (ud->verbose) {
					fprintf(stderr, ".. no fix\n");
				}
				break;

			default:
				if (ud->verbose) {
					fprintf(stderr, "status == %llu\n", gpsdata->set & MODE_SET);
				}
				break;
		}
	}

#if 0
#if GPSD_API_MAJOR_VERSION >= 7
	char *bp;
	if ((bp = strchr(gpsmessage, '\r')) != NULL)
		*bp = 0;
	fprintf(stderr, "(%zu) %s\n", gpsmessagelen, gpsmessage);
#endif
#endif

	if (valid == false)
		return;

	int_time = fix->time;

	if ((int_time == old_int_time) || fix->mode < MODE_2D) {
		// puts("rubbish");
		usleep(SIESTA);
		return;
	}

	/* may not be worth logging if we've moved only a very short distance */
	if (ud->displacement>0 && !first && earth_distance(
					fix->latitude,
					fix->longitude,
					old_lat, old_lon) < ud->displacement) {
		// puts("not enough move");
		usleep(SIESTA);
		return;
	}

	/* Don't log if interval seconds haven't elapsed since the last fix */
	if ((fabs(int_time - old_int_time) < ud->interval) && !first) {
		// puts("too soon");
		usleep(SIESTA);
		return;
	}

	if (first)
		first = false;

	old_int_time = int_time;
	if (ud->displacement > 0) {
		old_lat = fix->latitude;
		old_lon = fix->longitude;
	}

	print_fix(ud, gpsdata, int_time, PERIODIC_REPORT);
}

static int env_number(char *key, int min)
{
	char *p;
	int n;

	if ((p = getenv(key)) == NULL)
		return (min);
	if ((n = atoi(p)) < min)
		n = min;
	return (n);
}

void obsd_unveil(char *path, char *mode, char *comment)
{
#ifdef __OpenBSD__
	if (unveil(path, mode) != 0) {
		fprintf(stderr, "%s: unveil(%s, %s) on %s: %s\n",
			PROGNAME,
			path ? path : "NULL",
			mode ? mode : "NULL",
			comment,
			strerror(errno));
		exit(2);
	}
#endif
}

int main(int argc, char **argv)
{
	unsigned int flags = WATCH_NEWSTYLE; // WATCH_ENABLE | WATCH_JSON;
	int keepalive = 60, rc, c;
	char *p, *js;
	char *gpsd_host = "localhost", *gpsd_port = DEFAULT_GPSD_PORT;
	char *mqtt_host = "localhost";
	short mqtt_port = 1883;
	struct udata udata, *ud = &udata;
	char hostname[BUFSIZ], *h, *username;
	JsonNode *jo;
	char *cacert = NULL;

	ud->verbose = true;

	while ((c = getopt(argc, argv, "sv")) != EOF) {
		switch (c) {
			case 's':
				ud->verbose = false;
				break;
			case 'v':
				printf("ocli %s\n", VERSION);
				return (0);
				break;
			default:
				fprintf(stderr, "Usage: %s [-s] [-v] [parameter-file ..]\n", *argv);
				return (2);
		}
	}

#ifdef __OpenBSD__
	if (pledge("stdio inet unveil proc rpath exec", NULL) != 0) {
		perror("pledge");
		exit(2);
	}
#endif

	argc -= optind - 1;
	argv += optind - 1;

	ud->clientid = NULL;
	ud->interval = env_number("OCLI_INTERVAL", 1);		// minsecs seconds
	ud->displacement = env_number("OCLI_DISPLACEMENT", 0);	// minmove meters

	if ((p = getenv("GPSD_HOST")) != NULL)
		gpsd_host = strdup(p);

	if ((p = getenv("GPSD_PORT")) != NULL) {
		gpsd_port = atoi(p) < 1 ? DEFAULT_GPSD_PORT : p;
	}

	if ((p = getenv("MQTT_HOST")) != NULL)
		mqtt_host = strdup(p);

	if ((p = getenv("MQTT_PORT")) != NULL) {
		mqtt_port = atoi(p) < 1 ? 1883 : atoi(p);
	}

	if ((p = getenv("OCLI_CACERT")) != NULL) {
		cacert = strdup(p);
		obsd_unveil(cacert, "r", "$OCLI_CACERT");
	}

	if ((p = getenv("fixlog")) != NULL) {
		fixlog = fopen(p, "a");
		if (fixlog == NULL) {
			perror(p);
			exit(1);
		}
		obsd_unveil(p, "cw", "fixlog");
	}

	utarray_new(parms, &ut_str_icd);
	while (*++argv) {
		/* No need to unveil() each *argv because we have rpath promise */
		utarray_push_back(parms, &*argv);
	}

	obsd_unveil(NULL, NULL, "null veil");

	if ((username = getlogin()) == NULL)
		username = "nobody";

	if (gethostname(hostname, sizeof(hostname)) != 0)
		strcpy(hostname, "localhost");

	if ((h = strchr(hostname, '.')) != NULL)
		*h = 0;

	/*
	 * Build MQTT topic name which defaults to owntracks/user/device
	 * but can be overriden from the environment.
	 */

	if ((p = getenv("BASE_TOPIC")) != NULL) {
		ud->basetopic = strdup(p);
	} else {
		UT_string *to;

		utstring_new(to);
		utstring_printf(to, "owntracks/%s/%s", username, hostname);
		ud->basetopic = strdup(utstring_body(to));

		ud->username = strdup(username);
		ud->device =  strdup(hostname);

	}
	ud->tid = getenv("OCLI_TID");		/* may be null */

	if ((p = getenv("OCLI_CLIENTID")) != NULL) {
		ud->clientid = strdup(p);
	} else {
		UT_string *cid;

		utstring_new(cid);
		utstring_printf(cid, "ocli-%s-%s", username, hostname);
		ud->clientid = strdup(utstring_body(cid));
	}

	ud->t_cmd = malloc(strlen(ud->basetopic) + strlen("/cmd") + 1);
	sprintf(ud->t_cmd, "%s/cmd", ud->basetopic);

	ud->t_dump = malloc(strlen(ud->basetopic) + strlen("/dump") + 1);
	sprintf(ud->t_dump, "%s/dump", ud->basetopic);

	if (ud->verbose) {
		printf("t_base %s\n", ud->basetopic);
		printf("t_cmd  %s\n", ud->t_cmd);
		printf("t_dump %s\n", ud->t_dump );
	}

	mosquitto_lib_init();

	ud->mosq = mosquitto_new(ud->clientid, true, (void *)&udata);
	if (!ud->mosq) {
		fprintf(stderr, "Out of memory.\n");
		exit(1);
	}

	if (cacert != NULL) {
		rc = mosquitto_tls_set(ud->mosq,
			cacert,      /* cafile */
			NULL,        /* capath */
			NULL,        /* certfile */
			NULL,        /* keyfile */
			NULL         /* pw_callback() */
			);
		if (rc != MOSQ_ERR_SUCCESS) {
			fprintf(stderr, "%s: cannot set TLS CA: %s (check path names)\n",
				PROGNAME, mosquitto_strerror(rc));
			exit(3);
		}

		mosquitto_tls_opts_set(ud->mosq,
			SSL_VERIFY_PEER,
			NULL,                   /* tls_version: "tlsv1.2", "tlsv1" */
			NULL                    /* ciphers */
			);
	}



	/* Create payload for LWT consisting of starting timestamp */
	jo = json_mkobject();

	json_append_member(jo, "_type", json_mkstring("lwt"));
	json_append_member(jo, "tst", json_mknumber(time(0)));
	if ((js = json_stringify(jo, NULL)) != NULL) {
		if ((rc = mosquitto_will_set(ud->mosq, ud->basetopic, strlen(js), js, QOS1, true)) != MOSQ_ERR_SUCCESS) {
			fprintf(stderr, "Unable to set LWT: %s\n", mosquitto_strerror(rc));
		}
		free(js);
	}
	json_delete(jo);

	mosquitto_connect_callback_set(ud->mosq, cb_connect);
	mosquitto_disconnect_callback_set(ud->mosq, cb_disconnect);
	mosquitto_message_callback_set(ud->mosq, cb_message);

	if ((rc = mosquitto_connect(ud->mosq, mqtt_host, mqtt_port, keepalive)) != MOSQ_ERR_SUCCESS) {
		fprintf(stderr, "Unable to connect to %s:%d: %s\n", mqtt_host, mqtt_port,
			mosquitto_strerror(rc));
		perror("");
		exit(2);
	}

	signal(SIGINT, catcher);
	signal(SIGTERM, catcher);

	mosquitto_loop_start(ud->mosq);

	if (gps_open(gpsd_host, gpsd_port, &gpsdata) != 0) {
		fprintf(stderr,
		      "%s: no gpsd running or network error: %d, %s\n",
		      argv[0], errno, gps_errstr(errno));
		exit(EXIT_FAILURE);
	}

	gps_stream(&gpsdata, flags, NULL);

	while (1) {
		if (gps_waiting(&gpsdata, 5000000) == true) {
			errno = 0;
#if GPSD_API_MAJOR_VERSION >= 7
			if (gps_read(&gpsdata, gpsmessage, gpsmessagelen) == -1) {
#else
			if (gps_read(&gpsdata) == -1) {
#endif
				if (errno == 0) {
					// lost contact with gpsd
					break;
				}
			} else {
				conditionally_log_fix(&udata, &gpsdata);
			}
		}
	}

	gps_stream(&gpsdata, WATCH_DISABLE, NULL);
	gps_close(&gpsdata);

	utarray_free(parms);

	mosquitto_disconnect(ud->mosq);
	mosquitto_loop_stop(ud->mosq, false);
	mosquitto_lib_cleanup();

	exit(EXIT_SUCCESS);
}
